SnowplowTracker/SnowplowTracker/Emitter.cs

```diff
@@ -0,0 +1,490 @@
+			sending = true;
+			List<EventRow> events = eventStore.GetDescEventRange (sendLimit);
+
+			// If some events were found start sending!
+			if (events.Count > 0) {
+				Log.Debug ("Emitter: Event count: " + events.Count);
+				List<RequestResult> results = SendRequests (events);
+				events = null;
+
+				int success = 0;
+				int failure = 0;
+
+				foreach (RequestResult result in results) {
+					if (result.success) {
+						foreach (int id in result.rowIds) {
+							eventStore.DeleteEvent (id);
+						}
+						success += result.rowIds.Count;
+					} else {
+						failure += result.rowIds.Count;
+					}
+				}
+
+				Log.Debug (" + Successful: " + success);
+				Log.Debug (" + Failure: " + failure);
+
+				if (failure > 0 && success == 0) {
+					// All events failed; sleep for 10 seconds and then retry.
+					// - Sleep prevents thrashing of database in situations where nothing can send.
+					Log.Error ("Emitter: All events failed to send; pausing emitter for ten seconds.");
+					sending = false;
+					if (emitterTimer != null) {
+						emitterTimer.Change (FAIL_INTERVAL, Timeout.Infinite);
+					}
+				} else {
+					if (emitterTimer != null) {
+						Log.Debug ("Emitter: All events sent successfully; checking for more...");
+						Emit (null);
+					}
+				}
+			} else {
+				Log.Debug ("Emitter: No more events to send; waiting for more.");
+				sending = false;
+				lock (emitLock) {
+					Monitor.Wait(emitLock);
+				}
+				if (emitterTimer != null) {
+					Emit (null);
+				}
+			}
+		}
+
+		// --- Event Senders
+
+		/// <summary>
+		/// From a range of event rows will construct a list of RequestResult objects
+		/// </summary>
+		/// <returns>The results of sending all requests</returns>
+		/// <param name="eventRows">Event rows from the database</param>
+		private List<RequestResult> SendRequests(List<EventRow> eventRows) {
+			if (httpMethod == HttpMethod.GET) {
+				return HttpGet (eventRows);
+			} else {
+				return HttpPost (eventRows);
+			}
+		}
+
+		/// <summary>
+		/// Sends all events as GET requests on background threads
+		/// </summary>
+		/// <returns>The get.</returns>
+		/// <param name="eventRows">Event rows.</param>
+		private List<RequestResult> HttpGet(List<EventRow> eventRows) {
+			List<ReadyRequest> requests = new List<ReadyRequest>();
+			List<RequestResult> results = new List<RequestResult>();
+
+			// Send each row as an individual GET Request
+			foreach (EventRow eRow in eventRows) {
+				TrackerPayload payload = eRow.GetPayload();
+				long byteSize = payload.GetByteSize() + POST_STM_BYTES;
+				bool oversize = byteSize > byteLimitGet;
+				Log.Verbose("Emitter: Sending GET with byte-size: " + byteSize);
+				requests.Add(getGETRequest(payload.GetDictionary(), eRow.GetRowId(), oversize));
+			}
+
+			// Wait for the results of each request
+			foreach (ReadyRequest request in requests) {
+				request.WaitForResult();
+				results.Add (request.result);
+			}
+
+			requests = null;
```

Comment:
I believe that setting something to null is a big red flag for the GC to move it along from what I have read.  Otherwise it can get quite lazy and will only start releasing as it gets close to the memory cap, which can cause some issues on mobiles... Not sure if it will work the same in C# but it can't hurt!